File: lib/oereb.ts
/**
 * Cantonal ÖREB-Kataster webservices (public-law restrictions cadastre).
 * Every canton runs its own service implementing the standardized federal
 * interface; an official PDF extract is available at
 * {base}/extract/pdf?EGRID=<egrid>.
 *
 * Base URLs from the official directory: https://www.cadastre.ch/de/oereb-webservice
 * Cantons without a published standard service URL (e.g. BE, LU, NE, UR, VD, VS)
 * are omitted — the button is hidden for parcels there.
 */
const OEREB_SERVICES: Record<string, string> = {
  AG: 'https://api.geo.ag.ch/v2/oereb',
  AI: 'https://oereb.ai.ch/ktai/wsgi/oereb',
  AR: 'https://oereb.ar.ch/ktar/wsgi/oereb',
  BL: 'https://oereb.geo.bl.ch',
  BS: 'https://api.oereb.bs.ch',
  FR: 'https://maps.fr.ch/RDPPF_ws/RdppfSVC.svc',
  GE: 'https://ge.ch/terecadastrews/RdppfSVC.svc',
  GL: 'https://map.geo.gl.ch/oereb',
  GR: 'https://oereb.geo.gr.ch/oereb',
  JU: 'https://geo.jura.ch/crdppf_server',
  NW: 'https://oereb.gis-daten.ch/oereb',
  OW: 'https://oereb.gis-daten.ch/oereb',
  SG: 'https://oereb.geo.sg.ch/ktsg/wsgi/oereb',
  SH: 'https://oereb.geo.sh.ch',
  SO: 'https://geo.so.ch/api/oereb',
  SZ: 'https://map.geo.sz.ch/oereb',
  TG: 'https://map.geo.tg.ch/services/oereb',
  TI: 'https://crdpp.geo.ti.ch/oereb2',
  ZG: 'https://oereb.zg.ch/ors',
  ZH: 'https://maps.zh.ch/oereb/v2',
};

/**
 * URL of the official ÖREB PDF extract for a parcel, or null when the
 * canton has no published standard service or the EGRID is missing.
 */
export function oerebPdfUrl(canton: string, egrid: string): string | null {
  const base = OEREB_SERVICES[canton];
  if (!base || !egrid.startsWith('CH')) return null;
  return `${base}/extract/pdf?EGRID=${encodeURIComponent(egrid)}`;
}

/** Whether the canton publishes a standard ÖREB webservice. */
export function hasOerebService(canton: string): boolean {
  return canton in OEREB_SERVICES;
}

/** Picks the German text from an ÖREB multilingual `[{Language, Text}]` field. */
function germanText(value: unknown): string | null {
  if (typeof value === 'string') return value;
  if (Array.isArray(value)) {
    const entry =
      (value.find(
        (v) => typeof v === 'object' && v && (v as { Language?: string }).Language === 'de'
      ) as { Text?: string } | undefined) ?? (value[0] as { Text?: string } | undefined);
    return entry?.Text ?? null;
  }
  return null;
}

/** A precise cantonal land-use zone and the share of the parcel it covers. */
export interface ExactZone {
  zone: string;
  /** Percentage of the parcel in this zone, or null when no area is reported. */
  percent: number | null;
}

/** Formats zones as "80% Wohnzone 2, 20% Wohnzone 4" (or just the names). */
export function formatExactZones(zones: ExactZone[]): string {
  return zones
    .map((z) => (z.percent != null ? `${z.percent}% ${z.zone}` : z.zone))
    .join(', ');
}

/**
 * Fetches the precise cantonal land-use zone(s) for a parcel from the ÖREB
 * cadastre — e.g. "Wohnzone 2" instead of the harmonized "Wohnzonen" — together
 * with the share of the parcel each zone covers (from the ÖREB area shares).
 *
 * Returns the zones ordered by area share (largest first; a parcel can span
 * several zones), or null when the canton has no service, the request fails, or
 * no land-use plan is published for the parcel.
 */
export async function fetchExactZones(
  canton: string,
  egrid: string,
  signal?: AbortSignal
): Promise<ExactZone[] | null> {
  const base = OEREB_SERVICES[canton];
  if (!base || !egrid.startsWith('CH')) return null;
  try {
    const res = await fetch(
      `${base}/extract/json?GEOMETRY=false&EGRID=${encodeURIComponent(egrid)}`,
      { signal, headers: { Accept: 'application/json' } }
    );
    if (!res.ok) return null;
    const data = await res.json();

    // Recursively collect land-use (Nutzungsplanung) restrictions; structure
    // varies slightly between cantonal services, so we search rather than rely
    // on a fixed path.
    const found: { text: string; area: number; typeCode: string }[] = [];
    const visit = (node: unknown) => {
      if (Array.isArray(node)) {
        node.forEach(visit);
        return;
      }
      if (node && typeof node === 'object') {
        const obj = node as Record<string, unknown>;
        const themeCode = (obj.Theme as { Code?: string } | undefined)?.Code;
        if (
          typeof themeCode === 'string' &&
          themeCode.includes('Nutzungsplanung') &&
          obj.LegendText
        ) {
          const text = germanText(obj.LegendText);
          if (text) {
            found.push({
              text,
              area: typeof obj.AreaShare === 'number' ? obj.AreaShare : 0,
              typeCode: obj.TypeCode == null ? '' : String(obj.TypeCode),
            });
          }
        }
        Object.values(obj).forEach(visit);
      }
    };
    visit(data);

    // Keep only the Grundnutzung (the Zonenplan base zones) that actually cover
    // area. Two filters:
    //  - area > 0 drops the map-legend entries that carry no parcel share.
    //  - TypeCode 1–4 = Grundnutzung; 5–9 = overlay festlegungen (archäologische
    //    Zone, Ortsbildschutzzone, Gefahrenzone, Bebauungsplan, …) which are not
    //    part of the simple Zonenplan, so they are excluded.
    const hasAreas = found.some((f) => f.area > 0);
    const relevant = (hasAreas ? found.filter((f) => f.area > 0) : found).filter(
      (f) => !/^[5-9]/.test(f.typeCode)
    );

    // Sum area per actual zone designation (e.g. "Wohnzone 2", "Kernzone",
    // "Wohn- und Arbeitszone"), largest first.
    const byText = new Map<string, number>();
    for (const { text, area } of relevant) {
      byText.set(text, (byText.get(text) ?? 0) + area);
    }
    const sorted = Array.from(byText.entries()).sort((a, b) => b[1] - a[1]);
    const total = sorted.reduce((sum, [, area]) => sum + area, 0);

    const zones: ExactZone[] = sorted.map(([zone, area]) => ({
      zone,
      percent: hasAreas && total > 0 ? Math.round((area / total) * 100) : null,
    }));
    return zones.length > 0 ? zones : null;
  } catch {
    return null;
  }
}

File: backend/app/oereb.py
"""
OEREB (public-law restrictions cadastre) helpers.

Two things live here:

  * oereb_pdf_url   : build the link to a parcel's official OEREB PDF extract.
  * fetch_exact_zones: query a cantonal OEREB JSON extract and distil it down to
    the precise land-use zone(s) the parcel lies in (e.g. "Wohnzone 2" instead of
    the harmonized "Wohnzonen"), together with each zone's area share.

Every canton runs its own OEREB service (see config.OEREB_SERVICES). The JSON
structure varies slightly between cantons, so fetch_exact_zones searches the
whole document recursively for the relevant restriction objects rather than
relying on a fixed path.
"""

import re
from typing import Any, Dict, List, Optional
from urllib.parse import quote

import httpx

from .config import OEREB_SERVICES

# The OEREB JSON extract can be slow to generate, so it gets a longer timeout
# than the geo.admin.ch calls.
OEREB_TIMEOUT = 30.0


def oereb_pdf_url(canton: str, egrid: str) -> Optional[str]:
    """
    Build the URL of the official OEREB PDF extract for a parcel.

    Params:
      canton: two-letter canton code (e.g. "ZH").
      egrid:  the parcel's EGRID (must start with "CH" to be valid).

    Returns:
      The PDF extract URL, or None when the canton has no published service or
      the EGRID is not a real federal id.
    """
    base = OEREB_SERVICES.get(canton)
    if not base or not egrid.startswith("CH"):
        return None
    # url-encode the EGRID so it is safe as a query parameter value.
    return base + "/extract/pdf?EGRID=" + quote(egrid, safe="")


def _german_text(value: Any) -> Optional[str]:
    """
    Pick the German text out of an OEREB multilingual field.

    OEREB's LegendText (and similar fields) is either a plain string or a list of
    {"Language": ..., "Text": ...} entries. We return the string as-is, or the
    Text of the entry whose Language is "de", or failing that the first entry's
    Text.

    Params:
      value: a string, or a list of {Language, Text} dicts.

    Returns:
      The German (or fallback) text, or None when nothing usable is present.
    """
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        entry: Optional[Dict[str, Any]] = None
        # Prefer the explicit German entry.
        for v in value:
            if isinstance(v, dict) and v.get("Language") == "de":
                entry = v
                break
        # Otherwise fall back to the first entry.
        if entry is None and value:
            first = value[0]
            entry = first if isinstance(first, dict) else None
        if entry is not None:
            return entry.get("Text")
    return None


def _collect_nutzungsplanung(node: Any, found: List[Dict[str, Any]]) -> None:
    """
    Recursively collect Nutzungsplanung (land-use plan) restriction entries.

    Walks the entire decoded JSON (dicts and lists). Whenever it meets an object
    whose Theme.Code contains the substring "Nutzungsplanung" and that has a
    truthy LegendText, it records a {text, area, type_code} entry. It always
    recurses into every dict value and list item so that nested structures are
    fully searched regardless of the canton's exact layout.

    Params:
      node:  the current JSON node (dict, list, or scalar).
      found: the accumulator list that collected entries are appended to.

    Returns:
      None (results are accumulated into `found`).
    """
    if isinstance(node, list):
        for item in node:
            _collect_nutzungsplanung(item, found)
        return
    if isinstance(node, dict):
        theme = node.get("Theme")
        theme_code = theme.get("Code") if isinstance(theme, dict) else None
        legend = node.get("LegendText")
        if (
            isinstance(theme_code, str)
            and "Nutzungsplanung" in theme_code
            and legend
        ):
            text = _german_text(legend)
            if text:
                area_share = node.get("AreaShare")
                found.append(
                    {
                        "text": text,
                        # AreaShare is the share of the parcel covered by this
                        # restriction; default to 0 when it is missing/non-numeric.
                        "area": area_share
                        if isinstance(area_share, (int, float))
                        and not isinstance(area_share, bool)
                        else 0,
                        # TypeCode distinguishes Grundnutzung (1-4) from overlays (5-9).
                        "type_code": ""
                        if node.get("TypeCode") is None
                        else str(node.get("TypeCode")),
                    }
                )
        # Recurse into every value so nested objects are searched too.
        for value in node.values():
            _collect_nutzungsplanung(value, found)


async def fetch_exact_zones(canton: str, egrid: str) -> Dict[str, Any]:
    """
    Fetch the precise cantonal land-use zone(s) for a parcel from OEREB.

    Returns the harmonized-but-precise zone designations (e.g. "Wohnzone 2") and
    each zone's percentage share of the parcel, ordered largest first. The result
    is shaped for the /api/exact-zone endpoint: {available, zones}.

    Availability and resilience:
      * available is False when the canton has no service or the EGRID is not a
        federal id ("CH..."); in that case zones is [].
      * any HTTP error or exception while fetching/parsing degrades to
        {available: True, zones: []} rather than raising.

    Filtering logic (mirrors the frontend):
      * Collect every Nutzungsplanung restriction with a LegendText.
      * If any entry reports an area, keep only those with area > 0 (drops legend
        entries that carry no parcel share); otherwise keep all entries.
      * Exclude overlay festlegungen whose TypeCode starts with 5-9
        (archaeological zone, Ortsbildschutzzone, Gefahrenzone, Bebauungsplan...),
        keeping only the Grundnutzung base zones (TypeCode 1-4).
      * Sum the area per zone designation, sort descending, and express each as a
        percentage of the total (or null when no areas are reported).

    Params:
      canton: two-letter canton code.
      egrid:  the parcel's EGRID.

    Returns:
      A dict {"available": bool, "zones": [{"zone": str, "percent": int|None}]}.
    """
    base = OEREB_SERVICES.get(canton)
    available = base is not None and egrid.startswith("CH")
    if not available:
        return {"available": False, "zones": []}

    # From here on the canton is "available"; any failure yields empty zones.
    try:
        url = "{}/extract/json?GEOMETRY=false&EGRID={}".format(
            base, quote(egrid, safe="")
        )
        async with httpx.AsyncClient(timeout=OEREB_TIMEOUT) as client:
            res = await client.get(url, headers={"Accept": "application/json"})
        if res.status_code != 200:
            return {"available": True, "zones": []}
        data = res.json()
    except Exception:
        # Network error, timeout, invalid JSON, etc. -> just report no zones.
        return {"available": True, "zones": []}

    # Recursively gather all Nutzungsplanung entries from the document.
    found: List[Dict[str, Any]] = []
    _collect_nutzungsplanung(data, found)

    # area > 0 drops map-legend rows that carry no parcel share; if no entry has
    # an area we keep them all (some services omit AreaShare entirely).
    has_areas = any(f["area"] > 0 for f in found)
    candidate = [f for f in found if f["area"] > 0] if has_areas else found

    # TypeCode 1-4 = Grundnutzung (base zones we want); 5-9 = overlays to exclude.
    relevant = [f for f in candidate if not re.match(r"^[5-9]", f["type_code"])]

    # Sum the area per distinct zone designation.
    by_text: Dict[str, float] = {}
    for entry in relevant:
        by_text[entry["text"]] = by_text.get(entry["text"], 0) + entry["area"]

    # Sort zones by total area, largest first.
    sorted_zones = sorted(by_text.items(), key=lambda kv: kv[1], reverse=True)
    total = sum(area for _, area in sorted_zones)

    zones = [
        {
            "zone": text,
            # Percent share, or None when no areas were reported at all.
            "percent": round(area / total * 100)
            if (has_areas and total > 0)
            else None,
        }
        for text, area in sorted_zones
    ]
    return {"available": True, "zones": zones}
